src/WTG.Analyzers/Analyzers/Var/VarAnalyzer.cs

```diff
@@ -42,42 +45,87 @@ static void CompilationStart(CompilationStartAnalysisContext context)
 
-			if (candidate == null || candidate.ValueSource.IsKind(SyntaxKind.DefaultLiteralExpression))
+			if (decl == null || decl.Type.IsVar || decl.Variables.Count != 1)
+			{
+				return;
+			}
+
+			var expression = decl.Variables[0].Initializer?.Value;
+
+			if (expression == null || expression.IsKind(SyntaxKind.DefaultLiteralExpression))
 			{
 				return;
 			}
 
 			var model = context.SemanticModel;
-			var typeSymbol = (ITypeSymbol?)model.GetSymbolInfo(candidate.Type, context.CancellationToken).Symbol;
-			ITypeSymbol? expressionType = model.GetTypeInfo(candidate.ValueSource, context.CancellationToken).Type;
+			var declaredType = model.GetTypeInfo(decl.Type, context.CancellationToken).Type;
+			var expressionTypeInfo = model.GetTypeInfo(expression, context.CancellationToken);
+			var expressionType = expressionTypeInfo.Type;
 
-			if (typeSymbol == null || expressionType == null)
+			if (declaredType == null || expressionType == null)
 			{
 				return;
 			}
 
-			if (candidate.ValueSource.IsKind(SyntaxKind.StackAllocArrayCreationExpression) && expressionType.TypeKind == TypeKind.Struct && expressionType.IsMatch("System.Span`1"))
+			if (declaredType.IsReferenceType && expressionTypeInfo.Nullability.FlowState == NullableFlowState.NotNull)
+			{
+				// Unfortunately, `declaredTypeInfo.Nullability.Annotation` provides wildly inconsistent/inacurate information (no better than random)
+				// so we need to try get the information ourselves.
+				if (decl.Type.IsKind(SyntaxKind.NullableType))
+				{
+					// Declared as nullable but initialized with a non-null value. It may be set to null later.
+					return;
+				}
+			}
+
+			if (expression.IsKind(SyntaxKind.StackAllocArrayCreationExpression) && expressionType.TypeKind == TypeKind.Struct && expressionType.IsMatch(WellKnownTypeNames.Span))
 			{
 				return;
 			}
 
-			if (candidate.Unwrap)
+			if (TypeEquals(expressionType, declaredType))
 			{
-				expressionType = EnumerableTypeUtils.GetElementType(expressionType);
+				context.ReportDiagnostic(Rules.CreateUseVarWherePossibleDiagnostic(decl.Type.GetLocation()));
 			}
+		}
 
-			if (TypeEquals(expressionType, typeSymbol))
+		static void AnalyzeForEach(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
 			{
-				context.ReportDiagnostic(Rules.CreateUseVarWherePossibleDiagnostic(candidate.Type.GetLocation()));
+				return;
+			}
+
+			var node = (ForEachStatementSyntax)context.Node;
+
+			if (node.Type.IsVar)
+			{
+				return;
+			}
+
+			var model = context.SemanticModel;
+			var declaredType = model.GetTypeInfo(node.Type, context.CancellationToken).Type;
+			var expressionType = model.GetTypeInfo(node.Expression, context.CancellationToken).Type;
+
+			if (declaredType == null || expressionType == null)
+			{
+				return;
+			}
+
+			expressionType = EnumerableTypeUtils.GetElementType(expressionType);
```

Comment:
Does this work with extension enumerators in C# 9?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/extension-getenumerator

Reply:
I haven't tested it (and test it in a unit test until we upgrade to Roslyn 4.0). The implementation uses `ITypeSymbol.GetMembers` which I'm pretty sure I have seen include extension methods ... so it _probably_ would work.